holoviews/operation/datashader.py

```diff
@@ -0,0 +1,208 @@
+                paths.append(df)
+            kdims += element.kdims
+            vdims = element.vdims
+        elif isinstance(obj, Element):
+            line = isinstance(obj, Curve)
+            paths.append(obj.dframe())
+        if line:
+            empty = paths[0][:1].copy()
+            empty.loc[0, :] = (np.NaN,) * empty.shape[1]
+            paths = [elem for path in paths for elem in (path, empty)][:-1]
+        df = pd.concat(paths).reset_index(drop=True)
+        if category and df[category].dtype.name != 'category':
+            df[category] = df[category].astype('category')
+        return x, y, Dataset(df, kdims=kdims, vdims=vdims), line
+
+
+    def _process(self, element, key=None):
+        agg_fn = self.p.aggregator
+        category = agg_fn.column if isinstance(agg_fn, ds.count_cat) else None
+        x, y, data, line = self.get_agg_data(element, category)
+
+        if self.p.x_range:
+            xstart, xend = self.p.x_range
+        else:
+            xstart, xend = data.range(x)
+        if self.p.y_range:
+            ystart, yend = self.p.y_range
+        else:
+            ystart, yend = data.range(y)
+
+        # Compute highest allowed sampling density
+        width, height = self.p.width, self.p.height
+        if self.x_sampling:
+            x_range = xend - xstart
+            width = int(min([(x_range/self.p.x_sampling), width]))
+        if self.y_sampling:
+            y_range = yend - ystart
+            height = int(min([(y_range/self.p.y_sampling), height]))
+
+        cvs = ds.Canvas(plot_width=width, plot_height=height,
+                        x_range=(xstart, xend), y_range=(ystart, yend))
+
+        if line:
+            return cvs.line(data, x, y, self.p.aggregator)
+        else:
+            return cvs.points(data, x, y, self.p.aggregator)
+            
+
+def uint32_to_uint8(img):
+    return np.flipud(img.view(dtype=np.uint8).reshape(img.shape + (4,)))
+
+
+class Shade(ElementOperation):
+
+    cmap = param.ClassSelector(class_=(Iterable, Callable))
+
+    how = param.ObjectSelector(default='eq_hist', objects=['linear', 'log', 'eq_hist'])
+
+    @classmethod
+    def concatenate(cls, overlay):
+        if not isinstance(overlay, NdOverlay):
+            raise ValueError('Only NdOverlays can be concatenated')
+        xarr = xr.concat([v.data.T for v in overlay.values()], dim=overlay.kdims[0].name)
+        params = dict(get_param_values(overlay.last), vdims=overlay.last.vdims,
+                      kdims=overlay.kdims+overlay.last.kdims)
+        return Dataset(xarr.T, **params)
+
+    def _process(self, element, key=None):
+        if isinstance(element, NdOverlay):
+            bounds = element.last.bounds
+            element = self.concatenate(element)
+        else:
+            bounds = element.bounds
+
+        array = element.data[element.vdims[0].name]
+        kdims = element.kdims
+        shade_opts = dict(how=self.p.how)
+        if element.ndims > 2:
```

Comment:
A comment would be helpful here to summarize what this fairly convoluted processing of colormaps is for.